### PR TITLE
Added CosineDistance in distance.py, also included tests

### PR DESCRIPTION
**Pull request by RicherMans**
I added Cosine ~~Distance~~ Similarity, even though I am quite confused if this is necessary, since #806 was not included. Nevertheless, I ran tests locally with no errors.

### CODE REVIEW

File: torch/nn/modules/distance.py [hunk 1]
```diff
@@ -1,6 +1,39 @@
...
+            x2 = x2.view(1, -1)
+        #
+        #	 x1 * x2	 w12
+        #	--------- = -----
+        #	|x1|*|x2|	w1*w2
+        #
+        # Nominator
+        w12 = torch.sum(x1 * x2, 1)
+        w1 = 1. / (torch.norm(x1, 2, 1) + self.eps)
+        w2 = 1. / (torch.norm(x2, 2, 1) + self.eps)
```
Comment (apaszke) [inline torch/nn/modules/distance.py]: I think it's better to add eps and divide only once - compute `w1` and `w2` as norms and make the final expression be `w12 / ((w1 * w2) + self.eps).

File: torch/nn/modules/distance.py [hunk 2]
```diff
@@ -1,6 +1,39 @@
...
+        if x1.dim() == 1 or x2.dim() == 1:
+            x1 = x1.view(1, -1)
+            x2 = x2.view(1, -1)
+        #
+        #	 x1 * x2	 w12
+        #	--------- = -----
+        #	|x1|*|x2|	w1*w2
+        #
+        # Nominator
+        w12 = torch.sum(x1 * x2, 1)
```
Comment (apaszke) [inline torch/nn/modules/distance.py]: This is a dot product. Replace it with `torch.dot`.

File: torch/nn/modules/distance.py [hunk 3]
```diff
@@ -1,6 +1,39 @@
...
+        assert len(x) == 2, "Input needs to be two vectors"
+        x1,x2 = x
+        if x1.dim() == 1 or x2.dim() == 1:
+            x1 = x1.view(1, -1)
+            x2 = x2.view(1, -1)
+        #
+        #	 x1 * x2	 w12
+        #	--------- = -----
+        #	|x1|*|x2|	w1*w2
+        #
```
Comment (apaszke) [inline torch/nn/modules/distance.py]: Can you remove the comment? The function is very simple and it takes up a lot of space.

File: torch/nn/modules/distance.py [hunk 4]
```diff
@@ -1,6 +1,39 @@
...
+from .module import Module
 # TODO: Cosine
-# TODO: CosineDistance - make sure lua's CosineDistance isn't actually cosine similarity
+
+
+class CosineDistance(Module):
+    """
+    Computes cosine similarity between two vectors x1,x2:
+
+          x1*x2
```
Comment (apaszke) [inline torch/nn/modules/distance.py]: You need to make it a monospace block to have it displayed correctly. I'd be for using the `:math:` tag and LaTeX.

File: torch/nn/modules/distance.py [hunk 5]
```diff
@@ -1,6 +1,39 @@
-
+import torch
+from .module import Module
 # TODO: Cosine
-# TODO: CosineDistance - make sure lua's CosineDistance isn't actually cosine similarity
+
+
+class CosineDistance(Module):
```
Comment (apaszke) [inline torch/nn/modules/distance.py]: Can you rename it to `CosineSimilarity`? It's not a distance - the closer the inputs, the larger the output. Distance would be `1 - similarity`.

File: torch/nn/modules/distance.py [hunk 6]
```diff
@@ -1,6 +1,39 @@
...
+    """
+
+    def __init__(self):
+        super(CosineDistance, self).__init__()
+        self.eps = 1e-12
+
+    def forward(self, x):
+        assert len(x) == 2, "Input needs to be two vectors"
+        x1,x2 = x
+        if x1.dim() == 1 or x2.dim() == 1:
```
Comment (apaszke) [inline torch/nn/modules/distance.py]: I'm not sure what's the motivation behind that. There's no need to use ops that care about the dimensions, you can multiply any tensors that have the same number of elements, and `torch.norm` can return you a single scalar too.

File: torch/nn/modules/distance.py (hunk 2 above)
Comment (RicherMans) [inline torch/nn/modules/distance.py]: I am just a bit unclear here, in this case, I assume that the input might also be of size `batch x dim`, therefore a `torch.dot` would return a single scalar, removing the batch dimension completely.
Or did I misunderstand something?

Comment (apaszke) [inline torch/nn/modules/distance.py]: Right, the docs are quite misleading. It doesn't compute it on two vectors, but on batch of vectors. Can you change that please?

File: torch/nn/modules/distance.py (hunk 6 above)
Comment (apaszke) [inline torch/nn/modules/distance.py]: We don't support non-batched inputs, so there's no need for this check.

File: torch/nn/modules/distance.py (hunk 2 above)
Comment (RicherMans) [inline torch/nn/modules/distance.py]: You mean the docs for `torch.dot` ( that it does not support batched input ) or that the forward function needs to be specified in non-batch fashion?

Comment (apaszke) [inline torch/nn/modules/distance.py]: I meant that the docs for the function you wrote said that it computes it for a pair of vectors, while it really should be a pair of matrices. Sorry for that.

**Comment by fmassa** [comment]
Also, can you put this in `functional`, and instantiate it in the `Modules`?

**Comment by RicherMans** [comment]
I am unfamiliar, which `functional` and `Modules` do you refer to?
Is it the `_functions` in `torch.nn`, or the `functional` in `torch.nn`. 
Moreover, I put it already into the `modules.__init__` class, is that the one you referring to?

**Comment by fmassa** [comment]
@RicherMans Ideally, every `Module` should have a `functional` counterpart implementation. See for example the implementation for [pixel shuffle in functional](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L477-L505) and the [corresponding implementation as a module](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/pixelshuffle.py)

**Comment by RicherMans** [comment]
@fmassa Thanks for the clarification!

File: torch/nn/functional.py [hunk 7]
```diff
@@ -531,3 +531,20 @@ def upsample_bilinear(input, size=None, scale_factor=None):
         scale_factor (int): multiplier for spatial size. Has to be an integer.
     """
     return _functions.thnn.UpsamplingBilinear2d(size, scale_factor)(input)
+
+def cosine_similarity(input, eps=1e-12):
```
Comment (apaszke) [inline torch/nn/functional.py]: I think it's nicer to accept two inputs in here. That's the convention we have in all other parts of the API

File: torch/nn/functional.py [hunk 8]
```diff
@@ -531,3 +531,20 @@ def upsample_bilinear(input, size=None, scale_factor=None):
...
+    """
+    Function: cosine_similarity
+    Summary: Returns cosine similarity between two vectors ( here batches of vectors )
+    Examples: F.cosine_similarity(Variable(torch.randn(2,5,7)))
+    Attributes:
```
Comment (apaszke) [inline torch/nn/functional.py]: `Arguments:`

File: torch/nn/functional.py [hunk 9]
```diff
@@ -531,3 +531,20 @@ def upsample_bilinear(input, size=None, scale_factor=None):
         scale_factor (int): multiplier for spatial size. Has to be an integer.
     """
     return _functions.thnn.UpsamplingBilinear2d(size, scale_factor)(input)
+
+def cosine_similarity(input, eps=1e-12):
+    """
+    Function: cosine_similarity
+    Summary: Returns cosine similarity between two vectors ( here batches of vectors )
```
Comment (apaszke) [inline torch/nn/functional.py]: Replace these two lines with this:
```
Returns cosine similarity between two batches of vectors.
```

**Review by apaszke — COMMENTED** [review]
Can you please update the docstring so that they're consistent with all other? We're using different conventions.

File: test/test_nn.py [hunk 10]
```diff
@@ -1800,6 +1800,10 @@ def add_test(test):
         test_cuda=False
     ),
     dict(
+        constructor=lambda: nn.CosineSimilarity(),
+        input_size=(2,5,7),
```
Comment (fmassa) [inline test/test_nn.py]: nit: could you add spaces between the elements? `(2, 5, 7)` (for the lint checker)

File: torch/nn/functional.py [hunk 11]
```diff
@@ -531,3 +531,22 @@ def upsample_bilinear(input, size=None, scale_factor=None):
         scale_factor (int): multiplier for spatial size. Has to be an integer.
     """
     return _functions.thnn.UpsamplingBilinear2d(size, scale_factor)(input)
+
+def cosine_similarity(x1,x2, eps=1e-12):
```
Comment (fmassa) [inline torch/nn/functional.py]: nit: space between `x1` and `x2`

File: torch/nn/functional.py [hunk 12]
```diff
@@ -531,3 +531,22 @@ def upsample_bilinear(input, size=None, scale_factor=None):
...
+        x2 (Variable): Tensor with size (batch,dim), identical to x1
+        eps  (int, default=1e-12): Epsilon to avoid division by zero
+
+    Examples:
+        >>> cs = nn.CosineSimilarity()
+        >>> input = autograd.Variable(torch.Tensor(9, 4, 4))
+        >>> output = cs(input)
+        >>> print(output)
+    """
+    w12 = torch.sum(x1*x2, 1)
```
Comment (fmassa) [inline torch/nn/functional.py]: nit: space between `x1` and `x2`

File: torch/nn/modules/distance.py [hunk 13]
```diff
@@ -1,6 +1,28 @@
...
+        Args: 
+            - **x** of size (2, batch, size): Tensor containing the two input batches x1,x2 = x
+    """
+
+    def __init__(self):
+        super(CosineSimilarity, self).__init__()
+        self.eps = 1e-12
+
+    def forward(self, x):
+        return F.cosine_similarity(x,self.eps)
```
Comment (fmassa) [inline torch/nn/modules/distance.py]: The function signature has changed, you should pass `x1` and `x2` instead.

File: test/test_nn.py [hunk 14]
```diff
@@ -1800,6 +1800,10 @@ def add_test(test):
         test_cuda=False
     ),
     dict(
+        constructor=lambda: nn.CosineSimilarity(),
```
Comment (apaszke) [inline test/test_nn.py]: `module_name='CosineSimilarity'`

**Review by apaszke — COMMENTED** [review]
Can you please change that one line in tests and rebase the PR? Looks good now.

File: test/test_nn.py (hunk 14 above)
Comment (RicherMans) [inline test/test_nn.py]: Right, uhh might I ask how to pass to the testing framework two parameters for the `forward` method ?
`input=(torch.rand(5,7),torch.rand(5,7))` does not work, nor does it specifying `input_size=(2,5,7)`.

Comment (RicherMans) [inline test/test_nn.py]: I'm sorry but I havent found any way to include two parameters at `forward`. Until then I cant merge the thing since the test will fail. :(

Comment (apaszke) [inline test/test_nn.py]: Can you please do the same thing as in #835 (just add a separate test case)?

**Comment by soumith** [comment]
@RicherMans this is good to go, except that you have not rebased the PR correctly. You are touching 64 files :)
Any chance you can fix this?

**Comment by RicherMans** [comment]
I'll try, don't know why the rebase did touch so many files

File: torch/nn/modules/__init__.py [hunk 15]
```diff
@@ -22,7 +22,7 @@
     RNNCell, LSTMCell, GRUCell
 from .pixelshuffle import PixelShuffle
 from .upsampling import UpsamplingNearest2d, UpsamplingBilinear2d
-from .distance import PairwiseDistance
+from .distance import PairwiseDistance, CoineSimilarity
```
Comment (fmassa) [inline torch/nn/modules/__init__.py]: There is a typo in here.

**Comment by apaszke** [comment]
I've sent an equivalent PR #1502.
